File: leetcode/46_Permutations.py
import unittest
from typing import List

class Solution:
    def _rec(self, n: int, nums: List[int], i: int , combinations: List[List[int]]):
        if i >= n:
            combinations.append(nums.copy())
        else:
            self._rec(n, nums, i + 1, combinations)
            for j in range(i):
                nums[j], nums[i] = nums[i], nums[j]
                self._rec(n, nums, i + 1, combinations)
                nums[j], nums[i] = nums[i], nums[j]

    def permute(self, nums: List[int]) -> List[List[int]]:
        combinations = []
        self._rec(len(nums), nums, 0, combinations)
        return combinations

class TestPermute(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().permute([1,2,3]), [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]])        

    def test_2(self):
        self.assertEqual(Solution().permute([0,1]), [[0,1],[1,0]])

    def test_3(self):
        self.assertEqual(Solution().permute([1]), [[1]])

File: leetcode/46_Permutations.rs
struct Solution {}

impl Solution {
    fn _rec(n: usize, nums: &mut Vec<i32>, i: usize, combinations: &mut Vec<Vec<i32>>) {
        if  i >= n {
            combinations.push(nums.clone())
        } else {
            Self::_rec(n, nums, i + 1, combinations);
            for j in (0..i).rev() {
                nums.swap(i, j);
                Self::_rec(n, nums, i + 1, combinations);
                nums.swap(i, j);
            }
        }
    }

    pub fn permute(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut combinations: Vec<Vec<i32>> = vec![];
        Self::_rec(nums.len(), &mut nums, 1, &mut combinations);
        combinations
    }
}

#[cfg(test)]
mod permute_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::permute(vec![1,2,3]), vec![vec![1,2,3], vec![1,3,2], vec![2,1,3], vec![2,3,1], vec![3,1,2], vec![3,2,1]]);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::permute(vec![0,1]), vec![vec![0,1], vec![1,0]]);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::permute(vec![1]), vec![vec![1]]);
    }
}